cis194/week8/stevemao/Party.hs

```diff
@@ -0,0 +1,49 @@
+{-# OPTIONS_GHC -Wall #-}
+module Party where
+import Employee
+import Data.Tree
+import Data.List
+import Text.Read
+
+glCons :: Employee -> GuestList -> GuestList
+glCons e@(Emp _ efun) (GL el glfun) = GL (el ++ [e]) (glfun + efun)
+
+instance Semigroup GuestList where
+  GL el1 f1 <> GL el2 f2 = GL (el1 <> el2) (f1 + f2)
+
+instance Monoid GuestList where
+  mempty = GL [] 0
+
+moreFun :: GuestList -> GuestList -> GuestList
+moreFun gl1 gl2
+  | compare gl1 gl2 == GT = gl1
+  | otherwise = gl2
+
+treeFold :: (a -> [b] -> b) -> Tree a -> b
+treeFold f (Node a forest) = f a . fmap (treeFold f) $ forest
+
+nextLevel :: Employee -> [(GuestList, GuestList)]
+  -> (GuestList, GuestList)
+nextLevel boss xs = foldr (\(gl1, gl2) (gl1', gl2') -> (glCons boss gl1 <> gl2', gl2 <> gl1')) (mempty, mempty) xs
+
+moreFunTuple :: (GuestList, GuestList) -> GuestList
+moreFunTuple = uncurry moreFun
```

Comment:
If we only use uncurried `moreFunTuple` once, we don't need to bind it to a name. If we have to, we can just use `moreFunTuple'`